lib/absinthe/middleware.ex

```diff
@@ -0,0 +1,264 @@
+defmodule Absinthe.Middleware do
+  @moduledoc """
+  Middleware enables custom resolution behaviour on a field.
+
+  All resolution happens through middleware. Even `resolve` functions are middleware,
+  as the `resolve` macro is just
+
+  ```
+  quote do
+    middleware Absinthe.Resolution, unquote(function_ast)
+  end
+  ```
+
+  Resolution happens by reducing a list of middleware onto an `%Absinthe.Resolution{}`
+  struct.
+
+  ## Example
+
+  ```
+  defmodule MyApp.Web.Authentication do
+    @behaviour Absinthe.Middleware
+
+    def call(resolution, _opts) do
+      case resolution.context do
+        %{current_user: _} ->
+          resolution
+        _ ->
+          resolution
+          |> Absinthe.Resolution.put_result({:error, "unauthenticated"})
+      end
+    end
+  end
+  ```
+
+  By specifying `@behaviour Absinthe.Middleware` the compiler will ensure that
+  we provide a `def call` callback. This function takes an `%Absinthe.Resolution{}`
+  struct and will also need to return one such struct.
+
+  On that struct there is a `context` key which holds the absinthe context. This
+  is generally where things like the current user are placed. For more information
+  on how the current user ends up in the context please see our full authentication
+  guide on the website.
+
+  Our `def call` function simply checks the context to see if there is a current
+  user. If there is, we pass the resolution onward. If there is not, we update
+  the resolution state to `:halt` and place an error result.
+
+  Middleware can be placed on a field in three different ways:
+
+  1) Using the `set_middleware/2` callback in your schema.
+  2) Using the `Absinthe.Schema.Notation.middleware/2` macro used inside a field definition
+  3) Returning a `{:middleware, SomeMiddleware, opts}` tuple from a resolution function.
+
+  ## The `set_middleware/2` callback.
+
+  `set_middleware/2` is a function callback on a schema. When you `use Absinthe.Schema`
+  a default implementation of this function is placed in your schema. It is passed
+  the an Absinthe.Type.Field struct, as well as the Absinthe.Type.Object struct
+  that the field is a part of. The middleware for a field exists as a list on
+  the `Field` struct under the `:middleware` key.
+
+  So for example if your schema contained:
+
+  ```
+  object :user do
+    field :name, :string
+    field :age, :integer
+  end
+
+  query do
+    field :lookup_user, :user do
+      resolve fn _, _ ->
+        {:ok, %{name: "Bob"}}
+      end
+    end
+  end
+
+  def set_middleware(field, object) do
+    # what is field?
+    # what is object?
```

Comment:
Are these two "what is X?" comments meant for the end user? As a reader I'm not sure what these comments are supposed to teach me.